core/src/main/java/org/apache/iceberg/SchemaUpdate.java

```diff
@@ -533,6 +537,34 @@ private static Schema applyChanges(
       }
     }
 
+    Map<Integer, List<Integer>> specToDeletes = Maps.newHashMap();
+    if (base != null) {
+      for (int fieldIdToDelete : deletes) {
+        for (PartitionSpec spec : base.specs()) {
+          if (spec.schema().findField(fieldIdToDelete) != null) {
+            List<Integer> deletesForSpec =
+                specToDeletes.computeIfAbsent(spec.specId(), k -> Lists.newArrayList());
+            deletesForSpec.add(fieldIdToDelete);
+            specToDeletes.put(spec.specId(), deletesForSpec);
+          }
+        }
```

Comment:
I think we only need to do this logic if the current snapshot is not null, so probably want to re-organize the logic around that.